### ISSUE TITLE
API access

### ISSUE THREAD

**Issue by mattjbr123** [NERC-CEH/dri_gridded_data#24]
One of the things that came out of the work package meeting on 22-10-2024 is that API access to data stored on the object storage is not explicitly included in the workflow diagram. 

From my perspective I see this as arising because there is some confusion over what 'API Access' is meaning in this context. The plan is that the data on the object store will be available publically from anywhere (firewalls local-to-the-user not withstanding), and to have code provided on the data catalogue page and/or a link to an analysis platform with such code, that would allow the user to essentially treat the entire dataset as if it was on their own local filesystem. 
Is this an API? Is something extra needed to make it an API if not? And do we actually need that something extra? 
These are the questions that need clarification to me. 
@mikebrown your input would be especially valuable here I think (and input very welcome from anyone else involved in the project too!)

**Comment by mattjbr123** [comment]
Some helpful thoughts from @fsamreen:

A few thoughts …

 

Option 1 - provide users with direct URLs to download/interact with the data stored in the S3 bucket.
Pro - it is easier to set up, faster access and might be quick to download but direct access can expose sensitive data or bucket structures.
Cons - difficult to enforce controlled access which might be needed due to various reasons (security, data moving cost, etc).
Option 2 - build a REST API that acts as an intermediary and users interact with the API – handle data requests, process them, and retrieve the data from S3.
Pros- more control over who can access what and how (even we could give controlled access to some datasets through authentication methods). We will have to run API server (management overhead).
Cons - additional development task to implement APIs. Integration with other services would be secure.
 
An important question here is – ‘Who are the users and how would they like to interact with the data’? followed by - What is a sustainable value-added solution without excessive cost, maintenance and unnecessary implementation overheads? We might end up offering access through various methods including direct access to S3 as well as APIs or even through DataLabs.

**Comment by mattjbr123** [comment]
Knowing the use cases and user stories, [listed here](https://github.com/NERC-CEH/dri_gridded_data/issues/30), should help inform the API question...

**Comment by mattjbr123** [comment]
Ultimately we decided we'd probably want a mix of Option 1 and Option 2

**Comment by mattjbr123** [comment]
How we choose to access the zarr data may affect the API design, as we want to wrap it up with whatever library we use. It's a bit of a complicated landscape at the moment, with lots of libraries that allow S3 access in slightly different ways and that integrate slightly differently with xarray.
My understanding is that there are two main ways of doing this.
Use FSSpec/S3FS (S3FS is part of FSSpec) with the URL of the dataset on S3:
- [Example 1](https://tutorial.xarray.dev/intermediate/remote_data/cmip6-cloud.html#reading-from-the-remote-zarr-storage) (albeit using gcsfs instead of s3fs, but the idea is the same), [similar example here too](https://gallery.pangeo.io/repos/pangeo-gallery/cmip6/basic_search_and_load.html#Load-Data)
- Example 2:
```
import s3fs  
fs = s3fs.S3FileSystem(anon=True)  
store = zarr.storage.FSStore('/zarr-demo/store', fs=fs)  
g = zarr.open_group(store)  
```
(from https://zarr.readthedocs.io/en/stable/tutorial.html#io-with-fsspec)
- [Example 3](https://datalab.datalabs.ceh.ac.uk/resource/hj/objstrtest/lab/tree/access_obj-str_s3fs-open_access.ipynb) in UKCEH DataLabs (private to UKCEH, @dolegi I will give you access)

Use an intake or STAC catalogue to store information about the dataset that can be used by xarray to read in the dataset 
- [Example 1](https://gallery.pangeo.io/repos/pangeo-data/pangeo-tutorial-gallery/intake.html)
- [Example 2](https://catalog.pangeo.io/browse/master/ocean/sea_surface_height/)
- [Example 3](https://datalab.datalabs.ceh.ac.uk/resource/hj/objstrtest/lab/tree/access_obj-str_intake-open_access.ipynb) (datalabs as above)

Also relevant: https://gallery.pangeo.io/repos/earthcube2020/ec20_abernathey_etal/cloud_storage.html

Taking a closer look at these examples it seems to me that all intake is doing is storing/finding the URLs and then passing them to xarray via S3FS anyway. I'm sure it can do a lot more than this, but at least superficially it would seem that these two methods are kinda the same under the hood.

**Comment by mattjbr123** [comment]
Some more about STAC catalogues. 
Chatting with @dwest77a has opened my eyes to [STAC](https://stacspec.org/en/) a bit more. 
My understanding is now that it is more or less a specifically-formated JSON file that describes/catalogues the datasets and the files in the datasets. This can then be read by an "API" or (python) package that knows how to deal with this specific formatting of a JSON file. It doesn't actually sound too dissimilar to [intake](https://pypi.org/project/intake/). IIRC there still needs to be some thought put into how xarray then picks up the actual data or references to it from the catalogue, but some of the examples ([1](https://gallery.pangeo.io/repos/pangeo-data/pangeo-tutorial-gallery/intake.html), [2](https://catalog.pangeo.io/browse/master/ocean/sea_surface_height/), [3](https://datalab.datalabs.ceh.ac.uk/resource/hj/objstrtest/lab/tree/access_obj-str_intake-open_access.ipynb), [4](https://gallery.pangeo.io/repos/earthcube2020/ec20_abernathey_etal/cloud_storage.html)) in the previous comment should hopefully answer that.

@dwest77a has developed [DataPoint](https://cedadev.github.io/datapoint/), a STAC API (an API that can read STAC-formatted JSON files) originally designed around the STAC catalogues that CEDA hold that describe their huge catalogue of data, specifically the data that has been "[kerchunked](https://github.com/fsspec/kerchunk)" to allow for open access over the cloud. DataPoint also has the capability to "automatically ... open cloud datasets given the configuration information in the STAC records that are searched" which is something that could be really useful for our project here. I feel like DataPoint is close to being the API we are looking for. 
@dolegi could you take a look too and see what you think?

**Comment by dolegi** [comment]
Had a bit of a look into this, don't think I can add more than what you've already mentioned. I also came across [intake-esm](https://intake-esm.readthedocs.io/en/stable/index.html) which looks to be an alternative to STAC.


A reasonable next step to me looks to be to generate some STAC files for some of the resulting zarr files and try to get [DataPointMapper](https://cedadev.github.io/datapoint/mappers.html) to work with it.

It would also be interesting to get thoughts on this from @rodscott , maybe the catalogue is doing something related for some datasets already?

**Comment by dolegi** [comment]
We are looking at using the sensor things api for the timeseries data. This looks interesting https://sta2stac.readthedocs.io/en/latest/

**Comment by mattjbr123** [comment]
Thinking about this again (also see issue #34 for examples of what STAC implementation might look like)

Is there any advantage to having an API for programmatic access to the datasets vs what we currently (aim to) do which is provide (on the EIDC catalogue pages) an s3 link and code snippet for opening the dataset programmatically, with xarray able to show a lot of the metadata, and the EIDC catalogue the rest?

Are there any other advantages, such as GUIs/explorers that could be more easily built on top of a STAC API, that might be useful/advantageous for the gridded timeseries data product? 

Outside of the gridded timeseries datasets scope, what are the advantages of a data catalogue like the EIDC having a STAC API, vs perhaps the other options out there (e.g. geocroissant). @philipkershaw @dwest77a might you be able to help here? I remember vaguely we talked about STAC when we last met, but not the details!

What relevant developments (if any) are currently going on in EIDC development that the gridded timeseries data product could take advantage of (or vice versa - we could trial things under this FDRI product banner and see what works and could be taken up by the wider catalogue). @phtrceh are you able to advise here?

**Comment by philipkershaw** [comment]
Hi @mattjbr123! I would say use of STAC has enabled us to leverage a lot of community support and consequently made things easier for us - e.g. Python client and server-side reference implementations and working out standards. In the case of the latter, it has helped have more eyes and scrutiny on any given problem. I can't comment on the relative place of geocroissant but from my limited impression, STAC is much more established, more mature, has a bigger uptake and broader use case it supports. I expect @dwest77a and @agstephens could better comment on where geocroissant is up to :)

**Comment by agstephens** [comment]
Hi @mattjbr123, To address your specific question:

> Are there any other advantages, such as GUIs/explorers that could be more easily built on top of a STAC API, that might be useful/advantageous for the gridded timeseries data product?

I think that we have all been searching for the one solution that unites everything in this space (for many years). Whilst STAC is a useful cataloguing standard, with APIs, extensions, and many client tools (including GUIs) - it doesn't give us a single unified solution to how we deliver and access the _actual data_. 

Positives for STAC include:
- large uptake by the wider environmental/geospatial community
- expanding range of STAC-aware tools/services
- pre-existing examples of how to expose different data profiles (like gridded time series)

However, there is another layer that is an essential part of true interoperability, which is the aggregation and data-access layer that converts a data URI into a Python/Xarray/PyTorch/Tensorflow/Numpy object. I think that extra layer should be a significant part of focus at this stage - and we have put forward the idea of the `loadeds` library as a common interface to build around. 

For me, it is the interface directly into the data that makes the real difference to the user, such as a Kerchunk aggregation file. It turns out that both STAC and Croissant can be a pointer to that interface - but neither of them address that level of interoperability directly - so there lies our challenge. :-)

**Comment by mattjbr123** [comment]
Gotcha, thanks Ag, really appreciate the detailed explanation :)

**Comment by mattjbr123** [comment]
At the moment the approach we're taking in the aggregation & data-access layer is to provide a URL to the bucket that xarray can read in using fsspec and the like. (The data in the bucket at the moment is a 'one-size-fits-all/none' chunk size zarr store. The chunksize problem we may revisit, but for us using Zarr directly and not Kerchunk/VirtualiZarr makes sense as (to greatly summarise) we have to rechunk data anyway.)

A further question this has prompted is therefore whether further exploration into this approach (providing a URL and an Xarray/fsspec code snippet) would be advantageous - are there performance gains from other approaches for example. I shall look into ```loadeds``` - that one is new to me.

**Comment by mattjbr123** [comment]
As for the original question around STAC, it seems like it has merits for enabling useful services that can plug in to it's API, which might be nice, but whether or not we're likely to implement it is probably at the EIDC catalogue level and something @phtrceh is better placed to have opinions on!

**Comment by agstephens** [comment]
Hi @mattjbr123, `loadeds` is something that fell out of the end of the HIGH5/FRAME-FM projects. It just seems wise for us to build an EDS-wide toolkit for talking to our common formats and datasets. 

On the subject of providing a single URL to point to a dataset, that can be read by a _thing_ (such as `loadeds`) into `xarray` makes a lot of sense to me.

**Comment by mattjbr123** [comment]
Thanks Ag, sounds good! I'll keep an eye on ```loadeds```
